File: create_community_files.js
process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';

var fs = require('fs');
var deepExtend = require('extend');
var stringify = require('json-stable-stringify');
var winston = require('winston');
var moment = require('moment');
var restClient = require('node-rest-client').Client;

var options = {
	netmonUrl: "https://netmon.freifunk-franken.de/api/rest/routerlist/?limit=2000",
	communitysFile: 'communitys_franken.json',
	outputDir: '../freifunkfranken-community/',
	outputPrefix: '',
	cacheRouterlistFile: './cache.routerlist.json',
	cacheMaxAge: 30, //in minutes
};

var client = new restClient();
var routerlist = null;
var routers = null;

// gets loaded from file
var metacommunity = null;
var communities = null;

function extend(destination, source) {
	for ( var property in source) {
		if (!destination[property]) {
			destination[property] = source[property];
		} else if (destination[property] instanceof Array && source[property] instanceof Array) {
			Array.prototype.push.apply(destination[property], source[property]);
		} else {
			if (typeof destination[property] == 'object') {
				deepExtend(true, destination[property], source[property]);
			}
		}
	}
	return destination;
}

function saveCommunityData(data, output_filename) {
	fs.writeFile(output_filename, stringify(data, {space: 4}), function(err) {
		if (err) {
			winston.error(err);
		} else {
			winston.info("Updated " + output_filename);
		}
	});
}

function updateCommunity(community, onlineCount, offlineCount) {
	var output_filename = options.outputDir + options.outputPrefix + community.id + ".json";
	if (!community.data.state) {
		community.data.state = {};
	}
	community.data.state.nodes = onlineCount;
	community.data.state.lastchange = new Date().toJSON();
	
	winston.info("==============================================================");
	winston.info("Online nodes in community '" + community.data.name + "': " + onlineCount + " (offine: " + offlineCount + ") \n");
	saveCommunityData(community.data, output_filename);
}

function getDistance(p1, p2) {
	function toRad(n) {
		return n * (Math.PI / 180);
	}
	var dLat = toRad(p2.lat - p1.lat);
	var dLon = toRad(p2.lon - p1.lon);
	var lat1 = toRad(p1.lat);
	var lat2 = toRad(p2.lat);
	var a = Math.sin(dLat / 2) * Math.sin(dLat / 2) + Math.sin(dLon / 2) * Math.sin(dLon / 2) * Math.cos(p1.lat) * Math.cos(p2.lat);
	var c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
	var R = 6371; // km
	var d = R * c;
	return d;
}

function updateCommunities(data) {
	var onlineCount = 0, offlineCount = 0;
	routerlist = data.netmon_response.routerlist;
	routers = data.netmon_response.routerlist[0].router;
	
	// write community files
	var length = communities.length;
	var community = {};
	for (var i = 0; i < length; i = i + 1) {
		onlineCount = offlineCount = 0;
		winston.info("Routers in " + communities[i].data.name + " (radius: " + communities[i].radius + " km): ");
		winston.info("-------------------------------------------------------------");
		for (var j = 0; j < routers.length; j = j + 1) {
			var distance = getDistance({
				"lat" : parseFloat(routers[j].latitude),
				"lon" : parseFloat(routers[j].longitude)
			}, communities[i].data.location).toFixed(2);
			
			if (distance < communities[i].radius) {
				if (routers[j].statusdata[0].status[0] === "online") {
					winston.info("Node ON  " + routers[j].hostname[0] + ", distance: " + distance + " km");
					onlineCount = onlineCount + 1;
				} else {
					winston.info("Node OFF " + routers[j].hostname[0] + ", distance: " + distance + " km");
					offlineCount = offlineCount + 1;
				}
			}
		}
		community.id = communities[i].id;
		community.data = extend(communities[i].data, metacommunity.data);
		
		updateCommunity(community, onlineCount, offlineCount);
	}
}

function exec(instructions) {
	var terminal = require('child_process').spawn('bash');
	terminal.stdout.on('data', function (data) {
		console.log("" + data);
	});
	terminal.on('exit', function (code) {
		console.log("-------------------------------------------------------");
		console.log("DONE");
	});
	
	for (var i=0; i<instructions.length; i=i+1) {
		console.log('\n' + instructions[i]);
		console.log("-------------------------------------------------------");
		terminal.stdin.write(instructions[i] + "\n");
	}
	terminal.stdin.end();
}

function run() {
	fs.readFile(options.communitysFile, 'utf8', function read(err, data) {
		if (err) {
			throw err;
		}
		data = JSON.parse(data);
		metacommunity = data.metacommunity;
		communities = data.communities;
		
		fs.stat(options.cacheRouterlistFile, function(err, stats) {
			var minutesDiff = -1;
			if (!err) {
				var mtime = stats.mtime;
				var fileDate = moment(mtime);
				var now = moment();
				minutesDiff = now.diff(fileDate, 'minutes');
				winston.info("Age of local routerlist: " + minutesDiff + " minutes\n");
			}
			if (err || (minutesDiff > options.cacheMaxAge)) {
				winston.info("Loading routerlist from netmon...");
				//TODO split into multiple requests
				client.get(options.netmonUrl,
					function(data, response) {
						winston.info("saving data to file " + options.cacheRouterlistFile + "\n");
						fs.writeFile(options.cacheRouterlistFile, stringify(data, {space: 4}), function(err) {
							if (err) {
								winston.error(err);
							}
						});
						updateCommunities(data);
					}
				);
			} else  {
				winston.info("Loading local routerlist...");
				fs.readFile(options.cacheRouterlistFile, 'utf8', function read(err, data) {
					if (err) {
						throw err;
					}
					updateCommunities(JSON.parse(data));
				});
			}
		});
	});
}

run();

File: create_community_files.py
#!/usr/bin/python3

# Copyright 2016, Steffen Pankratz.
# SPDX-License-Identifier: AGPL-3
# For license see LICENSE or http://spdx.org/licenses/AGPL-3.0

from math import sin, cos, sqrt, asin, radians

import datetime
import json
import os
import urllib.request
import sys


def get_json_data_from_url(url):
    return json.loads(urllib.request.urlopen(url).read().decode())


def get_json_data_from_file(file):
    with open(file, 'r') as json_file:
        return json.load(json_file)


def get_distance_in_km(point_x, point_y):
    lat_x = point_x['lat']
    lon_x = point_x['lon']
    lat_y = point_y['lat']
    lon_y = point_y['lon']
    dlon = lon_y - lon_x
    dlat = lat_y - lat_x
    a = sin(dlat / 2) ** 2 + cos(lat_x) * cos(lat_y) * sin(dlon / 2) ** 2
    c = 2 * asin(sqrt(a))
    radius = 6373.0  # approximate radius of earth in km
    return radius * c


def merge(a, b, path=None):
    if path is None:
        path = []
    for key in b:
        if key in a:
            if isinstance(a[key], dict) and isinstance(b[key], dict):
                merge(a[key], b[key], path + [str(key)])
            elif isinstance(a[key], list) and isinstance(b[key], list):
                a[key].extend(b[key])
            elif a[key] == b[key]:
                pass  # same value
            else:
                pass  # keep value from a
        else:
            a[key] = b[key]
    return a


def get_router_count_by_community(routers, community):
    router_count_by_community = 0
    point_y = dict()
    point_y['lat'] = radians(float(community['data']['location']['lat']))
    point_y['lon'] = radians(float(community['data']['location']['lon']))
    for router in routers:
        if get_distance_in_km(router, point_y) < community['radius']:
            router_count_by_community += 1
    return router_count_by_community


def get_online_routers(nodelist_data):
    routers = []
    for router in nodelist_data['nodes']:
        if router['status']['online'] and 'position' in router:
            routers.append(
                {'lat': radians(router['position']['lat']),
                 'lon': radians(router['position']['long'])})
    return routers


def update_community_files(nodelist_url, communities_file, output_dir):
    nodelist_data = get_json_data_from_url(nodelist_url)
    routers = get_online_routers(nodelist_data)
    community_data = get_json_data_from_file(communities_file)
    for community in community_data['communities']:
        merge(community, community_data['metacommunity'])
        community['data']['state']['nodes'] = get_router_count_by_community(
            routers, community)
        community['data']['state']['lastchange'] = datetime.datetime.utcnow() \
                                                           .isoformat() + 'Z'
        with open(os.path.join(output_dir,
                               community['id'] + '.json'), 'w') as json_file:
            json.dump(community['data'], json_file, indent=4, sort_keys=True,
                      ensure_ascii=False)


if __name__ == '__main__':
    if len(sys.argv) < 2:
        exit('Please specify a directory for the community files.')
    output_directory = sys.argv[1]
    if not os.path.exists(output_directory):
        os.mkdir(output_directory)
    update_community_files(
        'https://monitoring.freifunk-franken.de/api/nodelist',
        'communitys_franken.json', output_directory)
